File: ios/YamabikoChat/Data/Models/OpenRouterModels.swift
import Foundation

struct OpenRouterReasoningCapabilities: Codable, Sendable, Equatable {
    static let gatewayEfforts = ["max", "xhigh", "high", "medium", "low", "minimal", "none"]

    var supportedEfforts: [String]?
    var exposesEffortSelection: Bool
    var defaultEffort: String?
    var defaultEnabled: Bool?
    var supportsMaxTokens: Bool
    var mandatory: Bool

    init(
        supportedEfforts: [String]? = nil,
        exposesEffortSelection: Bool = false,
        defaultEffort: String? = nil,
        defaultEnabled: Bool? = nil,
        supportsMaxTokens: Bool = false,
        mandatory: Bool = false
    ) {
        self.supportedEfforts = supportedEfforts
        self.exposesEffortSelection = exposesEffortSelection
        self.defaultEffort = defaultEffort
        self.defaultEnabled = defaultEnabled
        self.supportsMaxTokens = supportsMaxTokens
        self.mandatory = mandatory
    }

    var selectableEfforts: [String] {
        guard exposesEffortSelection else { return [] }
        let source = supportedEfforts ?? Self.gatewayEfforts
        var seen: Set<String> = []
        return source
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() }
            .filter { effort in
                guard !effort.isEmpty, seen.insert(effort).inserted else { return false }
                return !mandatory || effort != "none"
            }
    }

    enum CodingKeys: String, CodingKey {
        case supportedEfforts = "supported_efforts"
        case defaultEffort = "default_effort"
        case defaultEnabled = "default_enabled"
        case supportsMaxTokens = "supports_max_tokens"
        case mandatory
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        exposesEffortSelection = container.contains(.supportedEfforts)
        supportedEfforts = try container.decodeIfPresent([String].self, forKey: .supportedEfforts)
        defaultEffort = try container.decodeIfPresent(String.self, forKey: .defaultEffort)
        defaultEnabled = try container.decodeIfPresent(Bool.self, forKey: .defaultEnabled)
        supportsMaxTokens = try container.decodeIfPresent(Bool.self, forKey: .supportsMaxTokens) ?? false
        mandatory = try container.decodeIfPresent(Bool.self, forKey: .mandatory) ?? false
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        if exposesEffortSelection {
            if let supportedEfforts {
                try container.encode(supportedEfforts, forKey: .supportedEfforts)
            } else {
                try container.encodeNil(forKey: .supportedEfforts)
            }
        }
        try container.encodeIfPresent(defaultEffort, forKey: .defaultEffort)
        try container.encodeIfPresent(defaultEnabled, forKey: .defaultEnabled)
        if supportsMaxTokens {
            try container.encode(true, forKey: .supportsMaxTokens)
        }
        if mandatory {
            try container.encode(true, forKey: .mandatory)
        }
    }
}

struct OpenRouterModel: Codable, Sendable {
    struct Pricing: Codable, Sendable, Equatable {
        var prompt: String?
        var completion: String?
        var request: String?
        var inputCacheRead: String? = nil

        enum CodingKeys: String, CodingKey {
            case prompt
            case completion
            case request
            case inputCacheRead = "input_cache_read"
        }
    }

    struct TopProvider: Codable, Sendable {
        var availableProviders: [String]?
        var availableQuantizations: [String]?
        var maxCompletionTokens: Int?

        enum CodingKeys: String, CodingKey {
            case availableProviders = "available_providers"
            case availableQuantizations = "available_quantizations"
            case maxCompletionTokens = "max_completion_tokens"
        }
    }

    struct Architecture: Codable, Sendable {
        var inputModalities: [String]?
        var outputModalities: [String]?

        enum CodingKeys: String, CodingKey {
            case inputModalities = "input_modalities"
            case outputModalities = "output_modalities"
        }
    }

    var id: String
    var name: String
    var description: String?
    var pricing: Pricing
    var contextLength: Int?
    var topProvider: TopProvider?
    var reasoning: OpenRouterReasoningCapabilities? = nil
    var architecture: Architecture? = nil
    var supportedParameters: [String]? = nil

    enum CodingKeys: String, CodingKey {
        case id
        case name
        case description
        case pricing
        case contextLength = "context_length"
        case topProvider = "top_provider"
        case reasoning
        case architecture
        case supportedParameters = "supported_parameters"
    }
}

struct OpenRouterModelsResponse: Codable, Sendable {
    var data: [OpenRouterModel]
}

struct SimpleModel: Codable, Sendable, Equatable, Identifiable {
    var id: String
    var name: String
    var provider: String
    var topProvider: String?
    var contextLength: Int
    var promptPricePerMillion: Double
    var completionPricePerMillion: Double
    var isFree: Bool
    var availableProviders: [String]
    var availableQuantizations: [String]
    var reasoning: OpenRouterReasoningCapabilities? = nil
    var inputModalities: [String] = []
    var outputModalities: [String] = []
    var maxCompletionTokens: Int? = nil
    var supportsTools: Bool = false
    var supportsReasoning: Bool = false

    static func fromOpenRouterModel(_ model: OpenRouterModel) -> SimpleModel {
        func parse(_ values: String?...) -> Double {
            values.compactMap { Double($0 ?? "") }.first ?? 0
        }

        let promptPerToken = parse(model.pricing.prompt, model.pricing.completion, model.pricing.request)
        let completionPerToken = parse(model.pricing.completion, model.pricing.prompt, model.pricing.request)

        return SimpleModel(
            id: model.id,
            name: model.name,
            provider: model.id.split(separator: "/").first.map(String.init) ?? "unknown",
            topProvider: model.topProvider?.availableProviders?.first,
            contextLength: model.contextLength ?? 0,
            promptPricePerMillion: promptPerToken * 1_000_000,
            completionPricePerMillion: completionPerToken * 1_000_000,
            isFree: promptPerToken == 0 && completionPerToken == 0,
            availableProviders: model.topProvider?.availableProviders ?? [],
            availableQuantizations: model.topProvider?.availableQuantizations ?? [],
            reasoning: model.reasoning,
            inputModalities: model.architecture?.inputModalities ?? [],
            outputModalities: model.architecture?.outputModalities ?? [],
            maxCompletionTokens: model.topProvider?.maxCompletionTokens,
            supportsTools: model.supportedParameters?.contains("tools") == true,
            supportsReasoning: model.reasoning != nil || (model.supportedParameters ?? []).contains {
                ["include_reasoning", "reasoning", "reasoning_effort"].contains($0)
            }
        )
    }
}

extension SimpleModel {
    func matchesSearchQuery(_ query: String) -> Bool {
        let trimmedQuery = query.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmedQuery.isEmpty else { return true }

        let searchTerms = Self.normalizedSearchText(trimmedQuery).split(separator: " ")
        guard !searchTerms.isEmpty else { return false }

        let searchableFields = [id, name, provider].map(Self.normalizedSearchText)
        return searchTerms.allSatisfy { term in
            searchableFields.contains { field in
                field.contains(term) || field.replacingOccurrences(of: " ", with: "").contains(term)
            }
        }
    }

    private static func normalizedSearchText(_ value: String) -> String {
        value
            .folding(options: [.caseInsensitive, .diacriticInsensitive, .widthInsensitive], locale: .current)
            .unicodeScalars
            .map { CharacterSet.alphanumerics.contains($0) ? String($0) : " " }
            .joined()
    }
}

struct ModelEndpoint: Codable, Sendable, Equatable, Identifiable {
    var id: String { tag ?? "\(name)_\(providerName ?? "")_\(quantization ?? "")" }
    var name: String
    var contextLength: Double?
    var providerName: String?
    var tag: String?
    var quantization: String?
    var maxCompletionTokens: Double?
    var maxPromptTokens: Double?
    var supportedParameters: [String]?
    var status: Int?
    var pricing: OpenRouterModel.Pricing?

    enum CodingKeys: String, CodingKey {
        case name
        case contextLength = "context_length"
        case providerName = "provider_name"
        case tag
        case quantization
        case maxCompletionTokens = "max_completion_tokens"
        case maxPromptTokens = "max_prompt_tokens"
        case supportedParameters = "supported_parameters"
        case status
        case pricing
    }
}

struct OpenRouterEndpointOption: Sendable, Equatable, Identifiable {
    var id: String { tag }
    var tag: String
    var providerName: String
    var quantization: String?
    var supportedParameters: [String]
    var status: Int?
    var promptPricePerMillion: Double? = nil
    var cachedInputPricePerMillion: Double? = nil
    var completionPricePerMillion: Double? = nil
}

struct OpenRouterModelEndpointOptions: Sendable, Equatable {
    var modelId: String
    var endpoints: [ModelEndpoint]
    var providerEndpoints: [OpenRouterEndpointOption]
    var quantizations: [String]
}

struct ModelEndpointsEnvelope: Codable, Sendable {
    struct DataEnvelope: Codable, Sendable {
        var endpoints: [ModelEndpoint]
    }

    var data: DataEnvelope
}

struct ProviderInfo: Codable, Sendable, Equatable, Identifiable {
    var id: String { slug }
    var name: String
    var slug: String
}

struct ProvidersResponse: Codable, Sendable {
    var data: [ProviderInfo]
}

struct ProviderDirectory: Sendable, Equatable {
    var nameToSlug: [String: String]
    var slugToName: [String: String]

    static let empty = ProviderDirectory(nameToSlug: [:], slugToName: [:])

    static func fromList(_ list: [ProviderInfo]) -> ProviderDirectory {
        let nameToSlug = Dictionary(uniqueKeysWithValues: list.map { ($0.name.lowercased(), $0.slug) })
        let slugToName = Dictionary(uniqueKeysWithValues: list.map { ($0.slug, $0.name) })
        return ProviderDirectory(nameToSlug: nameToSlug, slugToName: slugToName)
    }

    func slugForName(_ name: String?) -> String? {
        guard let name else { return nil }
        return nameToSlug[name.lowercased()]
    }

    func nameForSlug(_ slug: String?) -> String? {
        guard let slug else { return nil }
        return slugToName[slug]
    }
}

File: app/src/main/java/com/porarri/yamabikochat/data/remote/OpenRouterModels.kt
package com.porarri.yamabikochat.data.remote

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.longOrNull

@Serializable
data class OpenRouterReasoningCapabilities(
    @SerialName("supported_efforts") val supportedEfforts: List<String>? = null,
    val exposesEffortSelection: Boolean = false,
    @SerialName("default_effort") val defaultEffort: String? = null,
    @SerialName("default_enabled") val defaultEnabled: Boolean? = null,
    @SerialName("supports_max_tokens") val supportsMaxTokens: Boolean = false,
    val mandatory: Boolean = false
) {
    val selectableEfforts: List<String>
        get() {
            if (!exposesEffortSelection) return emptyList()
            val source = supportedEfforts ?: GATEWAY_EFFORTS
            val seen = mutableSetOf<String>()
            return source
                .map { it.trim().lowercase() }
                .filter { effort ->
                    if (effort.isEmpty() || !seen.add(effort)) return@filter false
                    !mandatory || effort != "none"
                }
        }

    companion object {
        val GATEWAY_EFFORTS = listOf("max", "xhigh", "high", "medium", "low", "minimal", "none")
    }
}

@Serializable
data class OpenRouterModelsResponse(
    val data: List<OpenRouterModel>
)

@Serializable
data class OpenRouterModel(
    val id: String,
    val name: String,
    val description: String? = null,
    val pricing: ModelPricing,
    @SerialName("context_length") val contextLength: Int? = null,
    val architecture: ModelArchitecture? = null,
    @SerialName("top_provider") val topProvider: ModelProvider? = null,
    val created: Long? = null,
    val reasoning: OpenRouterReasoningCapabilities? = null
)

@Serializable
data class ModelPricing(
    val prompt: String? = null,
    val completion: String? = null,
    val request: String? = null,
    val image: String? = null,
    @SerialName("web_search") val webSearch: String? = null,
    @SerialName("internal_reasoning") val internalReasoning: String? = null
)

@Serializable
data class ModelArchitecture(
    val modality: String? = null,
    val tokenizer: String? = null,
    @SerialName("instruct_type") val instructType: String? = null
)

@Serializable
data class ModelProvider(
    @SerialName("max_completion_tokens") val maxCompletionTokens: Int? = null,
    @SerialName("is_moderated") val isModerated: Boolean? = null,
    @SerialName("supported_parameters") val supportedParameters: List<String>? = null,
    @SerialName("available_providers") val availableProviders: List<String>? = null,
    @SerialName("available_quantizations") val availableQuantizations: List<String>? = null
)

// UIで使用するためのシンプルなモデル情報
data class SimpleModel(
    val id: String,
    val name: String,
    val provider: String,
    val topProvider: String? = null,
    val contextLength: Int,
    // Stored in USD per 1M tokens for UI convenience
    val promptPricePerMillion: Double,
    val completionPricePerMillion: Double,
    val isFree: Boolean = false,
    val availableProviders: List<String> = emptyList(),
    val availableQuantizations: List<String> = emptyList(),
    val reasoning: OpenRouterReasoningCapabilities? = null
) {
    // 後方互換性のため、旧プロパティを維持（単位: USD/1 token）
    val promptPrice: Double
        get() = promptPricePerMillion / 1_000_000.0
    
    val completionPrice: Double
        get() = completionPricePerMillion / 1_000_000.0
    
    companion object {
        fun fromOpenRouterModel(model: OpenRouterModel): SimpleModel {
            // OpenRouter API pricing values are USD per 1 token (strings like "0.0000007")
            fun parsePrice(vararg candidates: String?): Double {
                return candidates
                    .asSequence()
                    .mapNotNull { it?.toDoubleOrNull() }
                    .firstOrNull() ?: 0.0
            }

            val promptPricePerToken = parsePrice(
                model.pricing.prompt,
                model.pricing.completion,
                model.pricing.request
            )
            val completionPricePerToken = parsePrice(
                model.pricing.completion,
                model.pricing.prompt,
                model.pricing.request
            )
            val provider = model.id.split("/").firstOrNull() ?: "unknown"
            // 1Mトークン単位に変換（×1,000,000）
            val promptPricePerMillion = promptPricePerToken * 1_000_000.0
            val completionPricePerMillion = completionPricePerToken * 1_000_000.0
            
            return SimpleModel(
                id = model.id,
                name = model.name,
                provider = provider,
                topProvider = model.topProvider?.availableProviders?.firstOrNull(),
                contextLength = model.contextLength ?: 0,
                promptPricePerMillion = promptPricePerMillion,
                completionPricePerMillion = completionPricePerMillion,
                isFree = promptPricePerToken == 0.0 && completionPricePerToken == 0.0,
                availableProviders = model.topProvider?.availableProviders ?: emptyList(),
                availableQuantizations = model.topProvider?.availableQuantizations ?: emptyList(),
                reasoning = model.reasoning
            )
        }
    }
}

@Serializable
data class ModelEndpointsEnvelope(
    val data: ModelEndpointsData
)

@Serializable
data class ModelEndpointsData(
    val id: String? = null,
    val name: String? = null,
    val description: String? = null,
    val architecture: ModelArchitecture? = null,
    val endpoints: List<ModelEndpoint> = emptyList()
)

@Serializable
data class ModelEndpoint(
    val name: String,
    @SerialName("context_length") val contextLength: Double? = null,
    val pricing: ModelPricing,
    @SerialName("provider_name") val providerName: String? = null,
    @SerialName("supported_parameters") val supportedParameters: List<String>? = null,
    val quantization: String? = null,
    @SerialName("max_completion_tokens") val maxCompletionTokens: Double? = null,
    @SerialName("max_prompt_tokens") val maxPromptTokens: Double? = null,
    val status: JsonElement? = null,
    @SerialName("uptime_last_30m") val uptimeLast30m: Double? = null,
    val tag: String? = null
)

// Providers directory response from /api/v1/providers
@Serializable
data class ProvidersResponse(
    val data: List<ProviderInfo>
)

@Serializable
data class ProviderInfo(
    val name: String,
    val slug: String
)

data class ProviderDirectory(
    val nameToSlug: Map<String, String>,
    val slugToName: Map<String, String>
) {
    companion object {
        fun fromList(list: List<ProviderInfo>): ProviderDirectory {
            val nameToSlug = list.associate { it.name.lowercase() to it.slug }
            val slugToName = list.associate { it.slug to it.name }
            return ProviderDirectory(nameToSlug, slugToName)
        }
        val EMPTY = ProviderDirectory(emptyMap(), emptyMap())
    }
    fun slugForName(name: String?): String? = name?.let { nameToSlug[it.lowercase()] }
    fun nameForSlug(slug: String?): String? = slug?.let { slugToName[it] }
}

// Helper to normalize status field (string | number | boolean → String?)
fun ModelEndpoint.statusText(): String? {
    val s = this.status
    if (s == null) return null
    return when (s) {
        is JsonPrimitive -> {
            when {
                s.isString -> s.content.ifBlank { null }
                s.booleanOrNull != null -> if (s.booleanOrNull == true) "up" else "down"
                s.longOrNull != null -> s.longOrNull?.toString()
                else -> s.toString()
            }
        }
        else -> s.toString()
    }
}

data class OpenRouterEndpointOption(
    val tag: String,
    val providerName: String,
    val quantization: String? = null,
    val supportedParameters: List<String> = emptyList(),
    val status: String? = null
) {
    val id: String get() = tag
}

data class OpenRouterModelEndpointOptions(
    val modelId: String,
    val endpoints: List<ModelEndpoint> = emptyList(),
    val providerEndpoints: List<OpenRouterEndpointOption> = emptyList(),
    val quantizations: List<String> = emptyList()
)
